PolyEngine/Core/Src/pe/core/storage/Queue.hpp

```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
 		/// <returns>True if present, false otherwise.</returns>
-		bool Contains(const T& rhs) const { return FindIdx(rhs) < GetSize(); }
+		bool contains(const T& rhs) const { return findIdx(rhs) < getSize(); }
+		
+		class Iterator : public BaseObject<>
+		{
+			friend class Queue<T>;
+		public:
+			using iterator_category = std::bidirectional_iterator_tag;
+			using value_type = T;
+			using difference_type = std::ptrdiff_t;
+			using pointer = T*;
+			using reference = T&;
+
+			bool operator==(const Iterator& rhs) const { return idx == rhs.idx; }
+			bool operator!=(const Iterator& rhs) const { return idx != rhs.idx; }
+			Iterator& operator=(const Iterator& rhs) = default;
+
+			T& operator*() const { return ref.m_data.at(idx); }
+
+			Iterator& operator++() { idx = ref.getNextIdx(idx); return *this; }
+			Iterator operator++(int) { Iterator ret(ref, idx); idx = ref.getNextIdx(idx); return ret; }
+			Iterator& operator--() { idx = ref.getNextIdx(idx); return *this; }
+			Iterator operator--(int) { Iterator ret(ref, idx); idx = ref.getNextIdx(idx); return ret; }
+		private:
+			Iterator(Queue<T>& queue, size_t index) : ref(queue), idx(index) {};
+			Queue<T>& ref;
+			size_t idx;
+		};
+
+		class ConstIterator : public BaseObject<>
+		{
+			friend class Queue<T>;
+		public:
+			using iterator_category = std::bidirectional_iterator_tag;
+			using value_type = T;
+			using difference_type = std::ptrdiff_t;
+			using pointer = T*;
+			using reference = T&;
+
+			bool operator==(const ConstIterator& rhs) const { return idx == rhs.idx; }
+			bool operator!=(const ConstIterator& rhs) const { return idx != rhs.idx; }
+
+			const T& operator*() const { return ref.m_data.at(idx); }
+
+			ConstIterator& operator++() { idx = ref.getNextIdx(idx); return *this; }
+			ConstIterator operator++(int) { ConstIterator ret(ref, idx); idx = ref.getNextIdx(idx); return ret; }
+			ConstIterator& operator--() { idx = ref.getNextIdx(idx); return *this; }
+			ConstIterator operator--(int) { ConstIterator ret(ref, idx); idx = ref.getNextIdx(idx); return ret; }
+		private:
+			ConstIterator(const Queue<T>& queue, size_t index) : ref(queue), idx(index) {};
+			const Queue<T>& ref;
+			size_t idx;
+
+		};
+
+		class ReverseIterator : public BaseObject<>
+		{
+			friend class Queue<T>;
+		public:
+			using iterator_category = std::bidirectional_iterator_tag;
+			using value_type = T;
+			using difference_type = std::ptrdiff_t;
+			using pointer = T*;
+			using reference = T&;
+
+			bool operator==(const ReverseIterator& rhs) const { return idx == rhs.idx; }
+			bool operator!=(const ReverseIterator& rhs) const { return idx != rhs.idx; }
+
+			T& operator*() const { return ref.m_data.at(idx); }
+
+			ReverseIterator& operator++() { increment(); return *this; }
+			ReverseIterator operator++(int) { ReverseIterator ret(ref, idx); increment(); return ret; }
+			ReverseIterator& operator--() { idx = ref.getPrevIdx(idx); return *this; }
```

Comment:
Since this is `ReverseIterator` `operator--` should call `getNextIdx`. Same for `ConstReverseIterator`